### PR TITLE
Memory leak + respect SRWebSocket codes

### PR DESCRIPTION
**Pull request by rex-remind101**
I found a memory leak in `PTURLRequestOperation` that was affecting our application when creating new `PTPusher` objects. Sometimes it would leak the entire pusher stack.

```objective-c
URLSession = [NSURLSession sessionWithConfiguration:sessionConfig delegate:self delegateQueue:nil];
```
Docs:

> **Important**
> The session object keeps a strong reference to the delegate until your app exits or explicitly invalidates the session. If you do not invalidate the session by calling the `invalidateAndCancel` or `finishTasksAndInvalidate` method, your app leaks memory until it exits.
> 

I solved this by making a lightweight proxy delegate object so that the rest of Pusher could safely deallocate when needed without regard to the session.

Additionally, I added `invalidateAndCancel` on `finish` in `PTURLRequestOperation`. If the session isn't invalidated the OS may never release it. Given that the operation is finished there's no reason to hold onto the session. Invalidating the session will eventually release the session and the delegate.

Additionally, we noticed that upon calling `disconnect` on `PTPusher` the `PTPusherConnection` would still attempt to reconnect indefinitely. This was due to `PTPusherConnection` not respecting the `SRStatusCode` codes handed back by `SRWebSocket`. I made a small change to `PTPusherConnection` to respect`SRStatusCodeNormal` and `SRStatusCodeGoingAway` which are returned on normal disconnection. I did not add any code to handle the remaining set of `SRStatusCode`s however, only the codes for normal disconnection.

### CODE REVIEW

File: Unit Tests/PTURLRequestOperationSpec.m [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+- (void)testFinishShouldInvalidateTheSessionDestroyingTheSessionDelegateOnDeallocation
+{
+  [self.subject finish];
+  self.subject = nil;
+  
+  [[NSRunLoop mainRunLoop] runUntilDate:[NSDate dateWithTimeIntervalSinceNow:1]];
+  
+  [[self.session should] beNil];
+  [[(NSObject *)self.sessionDelegate should] beNil];
```
Comment (rex-remind101) [inline Unit Tests/PTURLRequestOperationSpec.m]: Keep in mind, there is no safe way to test this using Kiwi. The block objects create a bunch of autorelease pools that won't give up the resources until they are complete.

**Comment by hamchapman** [comment]
Nice! All seems sensible to me. @lukeredpath do you have any thoughts?

**Comment by rex-remind101** [comment]
Sweet :) I'm trying to compile and run the Functional Specs that failed CI but I'm getting `'Constants.h' not found` on `#import "Constants.h"`. I don't see the file in the xcworkspace either, is there something I need to run to generate it?

**Comment by hamchapman** [comment]
It's a file that should be generated as part of the TravisCI process, based on this file:

https://github.com/pusher/libPusher/blob/master/Sample%20iOS/Constants.h.example

Here is the relevant part of the `.travis.yml` file.

https://github.com/pusher/libPusher/blob/master/.travis.yml#L7-L10

I've not got time right now to look at why this isn't working but if you can figure out the problem, that would obviously be great (if you have time!). 

Otherwise I'll get to this next week 👍

**Comment by rex-remind101** [comment]
@hamchapman you have any time to look at this?

**Comment by hamchapman** [comment]
Ah sorry, not yet @rex-remind101. I'll take a look today

**Comment by hamchapman** [comment]
@rex-remind101 please could you try rebasing on master and then pushing again? Thanks!

**Comment by rex-remind101** [comment]
@hamchapman did the thing :)

**Comment by hamchapman** [comment]
Hmmm strange that it's still failing at the same point. Going to checkout your branch and try running it all locally to see if I can reproduce not on Travis.

**Comment by hamchapman** [comment]
All seems to work fine locally... Going to keep on investigating

**Comment by hamchapman** [comment]
Ah I think it's because the env vars are set as secure and PRs from forks aren't allowed to have access to the env vars (security risk: https://github.com/travis-ci/travis-ci/issues/1500).

I've just created a new pull request using your code: https://github.com/pusher/libPusher/pull/250

**Pull request merged**